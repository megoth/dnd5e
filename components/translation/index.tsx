import React from "react";
import { Localized } from "@fluent/react";
import {
  getTranslationId,
  generateTranslationUrl,
} from "../../src/models/translation";

interface Props {
  id: string;
  vars?: Record<string, string>;
}

export default function Translation({ id, vars }: Props) {
  const url = generateTranslationUrl(id);
  return <Localized id={getTranslationId(url)} vars={vars} />;
}

Translation.defaultProps = {
  vars: {},
};
