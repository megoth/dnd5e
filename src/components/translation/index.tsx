import React from "react";
import { Localized } from "@fluent/react";
import { type FluentVariable } from "@fluent/bundle";

interface Props {
  id?: string;
  url?: string;
  vars?: Record<string, FluentVariable>;
}

export default function Translation({ id, url, vars = {} }: Props) {
  return (
    <Localized id={id} vars={vars}>
      <span>
        Failed Message: {id} [{url}]
      </span>
    </Localized>
  );
}
