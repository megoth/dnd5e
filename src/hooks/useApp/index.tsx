import { useContext } from "react";
import AppContext from "./context";

export default function useApp() {
  return useContext(AppContext);
}
