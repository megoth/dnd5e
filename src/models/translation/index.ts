import {
  asUrl,
  getSolidDataset,
  getStringWithLocale,
  getThingAll,
  ThingPersisted,
} from "@inrupt/solid-client";
import { skos } from "rdf-namespaces";
import { negotiateLanguages } from "@fluent/langneg";
import { FluentBundle, FluentResource } from "@fluent/bundle";
import NestedError from "nested-error-stacks";
import translationsTurtle from "../../../public/data/translations.ttl";
import { createLocalResponse } from "../../utils";
import { getErrorsDataset, generateErrorUrl } from "../error";

export const translationDatasetUrl =
  "http://localhost:3000/data/translations.ttl";

export function getDefaultBundle(bundles) {
  return bundles[0];
}

export function getTranslationId(url) {
  return url.replace(/(:|\/|\.|#)/g, "-");
}

export function generateTranslationUrl(id) {
  return `${translationDatasetUrl}#${id}`;
}

export function getFailedMessage(translationUrl) {
  return `[Translation for ${translationUrl} does not exist]`;
}

export function getMessage(bundle, id, args: Record<string, any> = {}) {
  const translationUrl = generateTranslationUrl(id);
  const translationId = getTranslationId(translationUrl);
  const message = bundle.getMessage(translationId);
  return message?.value
    ? bundle.formatPattern(message.value, args)
    : getFailedMessage(translationUrl);
}

export function getTranslations(datasets): Array<ThingPersisted> {
  return datasets.reduce(
    (translations, dataset) => translations.concat(getThingAll(dataset)),
    []
  );
}

export async function getTranslationsDataset() {
  return getSolidDataset(translationDatasetUrl, {
    fetch: () => Promise.resolve(createLocalResponse(translationsTurtle)),
  });
}

export async function getTranslationBundleAll(userLocales) {
  try {
    const [translationsDataset, errorsDataset] = await Promise.all([
      getTranslationsDataset(),
      getErrorsDataset(),
    ]);
    const translations = getTranslations([translationsDataset, errorsDataset]);
    const currentLocales = negotiateLanguages(userLocales, ["en-US"], {
      defaultLocale: "en-US",
    });
    return currentLocales.map((locale) => {
      const bundle = new FluentBundle(locale);
      translations.forEach((t) => {
        const value = getStringWithLocale(t, skos.definition, locale);
        if (!value) {
          return;
        }
        const id = getTranslationId(asUrl(t));
        bundle.addResource(new FluentResource(`${id} = ${value}`));
      });
      return bundle;
    });
  } catch (error) {
    throw new NestedError(generateErrorUrl("translationsLoadFailed"), error);
  }
}
