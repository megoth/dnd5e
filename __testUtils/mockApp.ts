import { AppModel, createApp, getBundleKey } from "../src/models/app";
import mockFluentBundle from "./mockFluentBundle";
import { appVocabURL } from "./mockAppIndexDataset";
import mockLanguage, { defaultLocale } from "./mockLanguage";
import mockResourceBundleMap from "./mockResourceBundleMap";
import mockAppIndex from "./mockAppIndex";

export const errorsURL = "https://example.com/global/errors.ttl";
export const faqsURL = "https://example.com/global/faqs.ttl";
export const translationsURL = "https://example.com/global/translations.ttl";
export const localizationsURL =
  "https://example.com/global/translations.en-US.ttl";

export function mockAppPropValue(val, bundleName = "global", locale = "en-US") {
  return bundleName
    ? {
        [getBundleKey(locale, bundleName)]: val,
      }
    : {
        [locale]: val,
      };
}

export default function mockApp(overrides: Partial<AppModel> = {}): AppModel {
  const currentLocale = overrides.currentLocale || defaultLocale;
  return {
    appVocabURL: overrides.appVocabURL || appVocabURL,
    currentLocale,
    fluentBundles: {
      [currentLocale]: mockFluentBundle({}, translationsURL, {
        locale: currentLocale,
      }),
      ...overrides.fluentBundles,
    },
    languages: [mockLanguage(currentLocale), ...(overrides.languages || [])],
    resourceBundles: {
      ...mockResourceBundleMap({
        label: "admin",
        locale: currentLocale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
      ...mockResourceBundleMap({
        locale: currentLocale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
      ...overrides.resourceBundles,
    },
  };
}

export function mockNorwegianApp(overrides: Partial<AppModel> = {}): AppModel {
  const locale = "nb-NO";
  return mockApp({
    currentLocale: locale,
    fluentBundles: {
      [locale]: mockFluentBundle({}, translationsURL, {
        locale,
      }),
    },
    languages: [
      mockLanguage(locale, {
        languageCode: "ðŸ‡³ðŸ‡´",
      }),
    ],
    resourceBundles: {
      ...mockResourceBundleMap({
        locale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
    },
    ...overrides,
  });
}

export function mockEmptyApp(): AppModel {
  return createApp(defaultLocale, appVocabURL, null);
}

export function mockUnloadedApp(): AppModel {
  return createApp(defaultLocale, appVocabURL, mockAppIndex());
}
