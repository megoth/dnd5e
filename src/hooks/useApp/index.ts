import { useContext, useEffect } from "react";
import { AppContext } from "../../contexts/app";
import { appIsLoadingLocalizations } from "../../models/app";

export default function useApp(bundles: string[] = []) {
  const { app, setBundles } = useContext(AppContext);
  useEffect(() => {
    if (!app || !appIsLoadingLocalizations(app, bundles)) {
      return;
    }
    setBundles(bundles);
  }, [app, bundles, setBundles]);
  return app;
}
