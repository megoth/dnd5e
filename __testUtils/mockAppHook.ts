import mockApp from "./mockApp";

export default function mockAppHook(mock: jest.Mock, app = mockApp()) {
  mock.mockReturnValue(app);
}
