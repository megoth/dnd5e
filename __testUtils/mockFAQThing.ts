import { mockThingFrom, setUrl } from "@inrupt/solid-client";
import { rdf } from "rdf-namespaces";
import { getAppTerm } from "../src/models/appIndex";
import { appVocabURL } from "./mockAppIndexDataset";
import { chain } from "../src/utils";
import { faqsURL } from "./mockApp";

export const faqId = "faq";
export const faqURL = `${faqsURL}#${faqId}`;
export const faqLabelURL = "https://example.com/#faqLabel";
export const faqDescriptionURL = "https://example.com/#faqDescription";

export default function mockFAQThing(id = faqId) {
  return chain(
    mockThingFrom(`${faqsURL}#${id}`),
    (t) => setUrl(t, rdf.type, getAppTerm("FAQ", { appVocabURL })),
    (t) => setUrl(t, getAppTerm("faqLabel", { appVocabURL }), faqLabelURL),
    (t) =>
      setUrl(
        t,
        getAppTerm("faqDescription", { appVocabURL }),
        faqDescriptionURL
      )
  );
}
