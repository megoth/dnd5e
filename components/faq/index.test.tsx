import React from "react";
import { render } from "@testing-library/react";
import useApp from "../../src/hooks/useApp";
import mockAppHook from "../../__testUtils/mockAppHook";
import FAQ, { TESTID_FAQ_DESCRIPTION, TESTID_FAQ_LABEL } from "./index";
import { getMessage } from "../../src/models/translation";
import mockFAQThing, {
  faqDescriptionURL,
  faqId,
  faqLabelURL,
} from "../../__testUtils/mockFAQThing";
import mockApp from "../../__testUtils/mockApp";

jest.mock("../../src/hooks/useApp");
const mockedAppHook = useApp as jest.Mock;

describe("FAQ", () => {
  const app = mockApp();

  beforeEach(() => mockAppHook(mockedAppHook, app));

  it("renders given a FAQ", () => {
    const faq = mockFAQThing();
    const { asFragment, getByTestId } = render(<FAQ faq={faq} />);
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTID_FAQ_LABEL).innerHTML).toEqual(
      getMessage(app, faqLabelURL)
    );
    expect(getByTestId(TESTID_FAQ_DESCRIPTION).firstChild.textContent).toEqual(
      getMessage(app, faqDescriptionURL)
    );
  });

  it("renders given an id", () => {
    const { asFragment, getByTestId } = render(<FAQ id={faqId} />);
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTID_FAQ_LABEL).innerHTML).toEqual(
      getMessage(app, faqLabelURL)
    );
    expect(getByTestId(TESTID_FAQ_DESCRIPTION).firstChild.textContent).toEqual(
      getMessage(app, faqDescriptionURL)
    );
  });
});
