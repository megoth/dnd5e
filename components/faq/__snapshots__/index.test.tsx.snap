// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FAQ renders given a FAQ 1`] = `
<DocumentFragment>
  <div
    class="content leading-normal max-w-prose mx-auto content"
  >
    <h2
      data-testid="faq-label"
    >
      [Translation for https://example.com/#faqLabel does not exist]
    </h2>
    <div
      data-testid="faq-description"
    >
      <p>
        [Translation for https://example.com/#faqDescription does not exist]
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FAQ renders given an id 1`] = `
<DocumentFragment>
  <div
    class="content leading-normal max-w-prose mx-auto content"
  >
    <h2
      data-testid="faq-label"
    >
      [Translation for https://example.com/#faqLabel does not exist]
    </h2>
    <div
      data-testid="faq-description"
    >
      <p>
        [Translation for https://example.com/#faqDescription does not exist]
      </p>
    </div>
  </div>
</DocumentFragment>
`;
